### PR TITLE
[SPARK-28593][CORE] Rename ShuffleClient to BlockStoreClient which more close to its usage

### PR DESCRIPTION
**Pull request by xuanyuanking**
## What changes were proposed in this pull request?

After SPARK-27677, the shuffle client not only handles the shuffle block but also responsible for local persist RDD blocks. For better code scalability and precise semantics(as the [discussion](https://github.com/apache/spark/pull/24892#discussion_r300173331)), here we did several changes:

- Rename ShuffleClient to BlockStoreClient. 
- Correspondingly rename the ExternalShuffleClient to ExternalBlockStoreClient, also change the server-side class from ExternalShuffleBlockHandler to ExternalBlockHandler. 
- Move MesosExternalBlockStoreClient to Mesos package.

Note, we still keep the name of BlockTransferService, because the `Service` contains both client and server, also the name of BlockTransferService is not referencing shuffle client only.

## How was this patch tested?

Existing UT.

### CODE REVIEW

**Comment by xuanyuanking** [comment]
@cloud-fan

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalBlockStoreClient.java [hunk 1]
```diff
@@ -46,8 +46,8 @@
  * BlockTransferService), which has the downside of losing the shuffle data if we lose the
  * executors.
  */
-public class ExternalShuffleClient extends ShuffleClient {
-  private static final Logger logger = LoggerFactory.getLogger(ExternalShuffleClient.class);
+public class ExternalBlockStoreClient extends BlockStoreClient {
```
Comment (cloud-fan) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalBlockStoreClient.java]: let's also update its classdoc

File: core/src/main/scala/org/apache/spark/network/BlockTransferService.scala [hunk 2]
```diff
@@ -26,24 +25,20 @@ import scala.reflect.ClassTag
...
 import org.apache.spark.internal.Logging
 import org.apache.spark.network.buffer.{FileSegmentManagedBuffer, ManagedBuffer, NioManagedBuffer}
-import org.apache.spark.network.shuffle.{BlockFetchingListener, DownloadFileManager, ShuffleClient}
+import org.apache.spark.network.shuffle.{BlockFetchingListener, BlockStoreClient, DownloadFileManager}
 import org.apache.spark.storage.{BlockId, EncryptedManagedBuffer, StorageLevel}
 import org.apache.spark.util.ThreadUtils
 
 private[spark]
-abstract class BlockTransferService extends ShuffleClient with Closeable with Logging {
+abstract class BlockTransferService extends BlockStoreClient with Logging {
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/network/BlockTransferService.scala]: shall we add classdoc and mention that this class contains both server and client?

File: core/src/main/scala/org/apache/spark/storage/BlockManager.scala [hunk 3]
```diff
@@ -130,7 +130,7 @@ private[spark] class BlockManager(
     shuffleManager: ShuffleManager,
     val blockTransferService: BlockTransferService,
     securityManager: SecurityManager,
-    externalShuffleClient: Option[ExternalShuffleClient])
+    externalShuffleClient: Option[ExternalBlockStoreClient])
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: rename the variable as well?

File: core/src/main/scala/org/apache/spark/storage/BlockManager.scala [hunk 4]
```diff
@@ -459,7 +459,7 @@ private[spark] class BlockManager(
     for (i <- 1 to MAX_ATTEMPTS) {
       try {
         // Synchronous and will throw an exception if we cannot connect.
-        shuffleClient.asInstanceOf[ExternalShuffleClient].registerWithShuffleServer(
+        shuffleClient.asInstanceOf[ExternalBlockStoreClient].registerWithShuffleServer(
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: nit: is this `asInstanceOf` necessary?

**Comment by cloud-fan** [comment]
the new name loos clearer, also cc @attilapiros @vanzin @tgravescs

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalBlockStoreClient.java (hunk 1 above)
Comment (xuanyuanking) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ExternalBlockStoreClient.java]: Ah, yep, done in ae64538.

File: core/src/main/scala/org/apache/spark/network/BlockTransferService.scala (hunk 2 above)
Comment (xuanyuanking) [reply core/src/main/scala/org/apache/spark/network/BlockTransferService.scala]: Copy that, done in ae64538.

File: core/src/main/scala/org/apache/spark/storage/BlockManager.scala (hunk 3 above)
Comment (xuanyuanking) [reply core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: Initially I tried this, but there are plenty of names bind with `Shuffle`, so in ae64538 I just change this param name and the member var `shuffleClient`.

File: core/src/main/scala/org/apache/spark/storage/BlockManager.scala (hunk 4 above)
Comment (xuanyuanking) [reply core/src/main/scala/org/apache/spark/storage/BlockManager.scala]: Yep it's necessary because only ExternalBlockStoreClient has this function.

**Comment by xuanyuanking** [comment]
retest this please.

**Comment by tgravescs** [comment]
I'm ok with renaming, it is more clear. My only concern is changing public api's when maybe not truly needed, although this is 3.0 so the time to do it and I don't think this will impact very many users.

**Comment by cloud-fan** [comment]
AFAIK the `network` is an internal module, these interfaces/classes are not even developer APIs. I think it's fine to change them.

**Review by attilapiros — COMMENTED** [review]
I think the new names are much bette even self explanatory. 

But what about going one step further and renaming the `ExternalShuffleService` to `ExternalBlockStoreService`?

I know it will generate a lot of documentation changes and a temporary mapping in the mind of the operators running Spark but if we would like to ever fix this then this is the best time to do so.

cc @squito

**Comment by xuanyuanking** [comment]
Thanks for the review and comments.
@attilapiros Yep, I tried to rename the `ExternalShuffleService` to `ExternalBlockStoreService`, compare with the present changes, it's more like an API change and influence for end users. After changing the class to `ExternalBlockStoreService`, we'll get a dilemma that should we also change the plenty of variable names, related classes and configs name. Maybe we just limit the scope to the developers, better not to the end-users. WDYT?

**Comment by cloud-fan** [comment]
It takes time to educate the users that external shuffle service is now external block service. We need to update documents, configs, etc. and probably need a few blog posts, presentations, etc.

I don't think it's a good idea to do them with one single PR. I'd like to merge this PR first, as a start of internal renaming.

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by attilapiros** [comment]
ok, I see skipping its rename was result of a conscientious decision. That's fine.

**Comment by xuanyuanking** [comment]
Yep, thanks for the advice and review :)

**Pull request closed without merging**